server/pluginapi/mutex_mock.go

```diff
@@ -0,0 +1,41 @@
+package pluginapi
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	ErrLockTimeout = errors.New("timeout")
+)
+
+type ClusterMutexMock struct {
```

Comment:
Is `Mock` the right term? This is a functional implementation of a mutex. It's just in memory and not cluster-wide. I wonder if a name like `MemoryMutex` would be more self-explaining.

Another question: It's only used in tests, right? Should it live in a `_test.go` file then?

Reply:
The mocks here and in MM server are not in `_test.go` files.  The only difference between this one and the others is that it is not auto-generated.

Reply:
This is still a mock.  Before mocks were auto-generated, this is typically how mocks were written - a bare-bones, incomplete implementation of something more complex to satisfy a test.  We could not use this outside of unit tests as it is a terrible implementation of a mutex.  IMO, renaming it MemoryMutex invites it to be used elsewhere.

Reply:
Good point on the name "mock". I forgot that bare-bone implementations are still called mocks.

I would still like to see the mock implementation tucked away in a separate `pluginapi_test` packet or a separate test file, but the name makes it clear that it should not be used in production.

Reply:
@hanzei I tried putting the mock into a `_test.go` file and it caused errors.  Test files cannot export to other packages.  This is why the mocks in `mock_pluginapi` are not in test files, and why shared [test utilities](https://github.com/mattermost/mattermost/blob/master/server/channels/api4/apitestlib.go) are not either.